polkadot/xcm/xcm-executor/src/lib.rs

```diff
@@ -444,26 +467,88 @@ impl<Config: config::Config> XcmExecutor<Config> {
 		Ok(())
 	}
 
-	fn take_fee(&mut self, fee: Assets, reason: FeeReason) -> XcmResult {
+	fn take_fee(&mut self, fees: Assets, reason: FeeReason) -> XcmResult {
 		if Config::FeeManager::is_waived(self.origin_ref(), reason.clone()) {
 			return Ok(())
 		}
 		log::trace!(
 			target: "xcm::fees",
-			"taking fee: {:?} from origin_ref: {:?} in fees_mode: {:?} for a reason: {:?}",
-			fee,
+			"taking fees: {:?} from origin_ref: {:?} in fees_mode: {:?} for a reason: {:?}",
+			fees,
 			self.origin_ref(),
 			self.fees_mode,
 			reason,
 		);
+		let first = fees.get(0).ok_or(XcmError::AssetNotFound)?;
+		// If `BuyExecution` was called, we know we can try to use that asset for fees.
+		log::trace!(target: "xcm", "Asset for fees: {:?}", self.asset_for_fees);
+		let actual_asset_to_use = if let Some(asset_for_fees) = &self.asset_for_fees {
+			match Config::AssetConverter::convert_asset(&first, &asset_for_fees) {
+				Ok(new_asset) => new_asset,
+				// If we can't convert, then we return the original asset.
+				// It will error later in any case.
+				Err(error) => {
+					log::error!(
+						target: "xcm::take_fee",
+						"Could not convert fees to {:?}. Error: {:?}",
+						asset_for_fees,
+						error,
+					);
+					first.clone()
+				},
+			}
+		} else {
+			// We try each item in holding.
+			// TODO: We don't need this if we had a way to tell the executor what asset we want to use
+			// for fees.
+			// Option 1: Always using a `BuyExecution` instruction, even locally where we would normally not use it.
+			// This has the downside of being a worse experience.
+			// Option 2: Having an instruction that sets the asset to be used for fees.
+			// Can only do it on a new version.
+			// Option 3: Change the entrypoint `prepare_and_execute` to take the fee item.
+			// Is a breaking change.
+			// Option 4: Create a new entrypoint `prepare_and_execute_with_fee_asset`.
+			// Should work and is not a breaking change.
+			// TODO: Might increase benchmarks too much. Should optimize for the average case.
+			let mut asset_for_fees = first.clone();
+			// Holding is empty when using jit withdrawal. Could remove it.
+			// It's not empty because of jit withdrawal, it's empty because it's taken from holding before sending.
+			// That's the problem jit withdrawal was solving, we need to "park" delivery fees in the same way we did
+			// for `DepositReserveAsset`.
+			for asset in self.holding.fungible_assets_iter() {
```

Comment:
This way we basically deciding for a user which asset to use for the payment, this can be a surprise.

Reply:
Yeah, but some API needs to change or I have to add a new one to not have to do this.
I was thinking of creating a new entrypoint `prepare_and_execute_with_fee_asset` that `pallet-xcm` can use.
That'd keep the API the same for all other teams.
What do you think?